File: hackerrank/problem solving/algorithms/2 - implementation/19 - the hurdle race/with morfo library/the_hurdle_race.go
// https://www.hackerrank.com/challenges/the-hurdle-race/problem?isFullScreen=true

package main

import (
    "fmt"
    "morfo/morfoio"
    "slices"
)

func main() {
    morfoio.SkipInput()
    maxJump := morfoio.Read[int]()
    hurdles := morfoio.ReadLn[int]()
    fmt.Println(hurdleRace(hurdles, maxJump))
}

// n: length of hurdles
// T: O(n)
// S: O(1) extra space
func hurdleRace(hurdles []int, maxJump int) int {
    return max(0, slices.Max(hurdles) - maxJump)
}

File: hackerrank/problem solving/algorithms/2 - implementation/19 - the hurdle race/with morfo library/the_hurdle_race.py
# https://www.hackerrank.com/challenges/the-hurdle-race/problem?isFullScreen=true

import morfo


def main():
    morfo.skip_input()
    max_jump = morfo.read(int)
    hurdles = morfo.readln(int, list)
    print(hurdle_race(hurdles, max_jump))


# n: length of hurdles
# T: O(n)
# S: O(1) extra space
def hurdle_race(hurdles, max_jump):
    return max(0, max(hurdles) - max_jump)


if __name__ == '__main__':
    main()
